File: leetcode/algorithms/cpp/Kth Largest Element in an Array/Kth Largest Element in an Array.cpp
#include <vector>
using namespace std;
class Solution {
public:
  int findKthLargest(vector<int> &nums, int k) {
    return quickSelect(nums, 0, nums.size() - 1, k);
  }
  int quickSelect(vector<int> &nums, int lo, int hi, int k) {
    if (lo >= hi) {
      return nums[lo];
    }
    int i = lo - 1, j = hi + 1, pivot = nums[lo];
    while (i < j) {
      do {
        i++;
      } while (nums[i] < pivot);
      do {
        j--;
      } while (nums[j] > pivot);
      if (i < j) {
        swap(nums[i], nums[j]);
      }
    }
    int cnt = hi - j;
    if (k <= cnt) {
      return quickSelect(nums, j + 1, hi, k);
    } else {
      return quickSelect(nums, lo, j, k - cnt);
    }
  }
};

File: leetcode/algorithms/Kth Largest Element in an Array/kth-largest-element-in-an-array.py
from typing import List
import random


class Solution:
    def findKthLargestV2(self, nums: List[int], k: int) -> int:
        pivot = random.choice(nums)
        left = [x for x in nums if x < pivot]
        mid = [x for x in nums if x == pivot]
        right = [x for x in nums if x > pivot]

        if k <= len(right):
            return self.findKthLargestV2(right, k)
        elif k > len(right) + len(mid):
            return self.findKthLargestV2(left, k - (len(right) + len(mid)))
        return mid[0]

    def findKthLargest(self, nums: List[int], k: int) -> int:
        def quickSelect(lo: int, hi: int, k: int):
            if lo >= hi:
                return nums[lo]
            x = nums[lo]
            i, j = lo - 1, hi + 1
            while i < j:
                i += 1
                while i <= j:
                    if nums[i] >= x:
                        break
                    i += 1
                j -= 1
                while j >= i:
                    if nums[j] <= x:
                        break
                    j -= 1
                if i < j:
                    nums[i], nums[j] = nums[j], nums[i]
            largeCnt = hi - j
            if k <= largeCnt:
                return quickSelect(j + 1, hi, k)
            else:
                return quickSelect(lo, j, k - largeCnt)

        return quickSelect(0, len(nums) - 1, k)


print(Solution().findKthLargest([3, 2, 1, 5, 6, 4], 2))
print(Solution().findKthLargest([3, 2, 3, 1, 2, 4, 5, 5, 6], 4))
